src/util/finalizable_processor.rs

```diff
@@ -140,6 +141,15 @@ pub struct Finalization<E: ProcessEdgesWork>(PhantomData<E>);
 
 impl<E: ProcessEdgesWork> GCWork<E::VM> for Finalization<E> {
     fn do_work(&mut self, worker: &mut GCWorker<E::VM>, mmtk: &'static MMTK<E::VM>) {
+        // Rescan soft and weak references at the end of the transitive closure from resurrected
```

Comment:
* Do we need this `ReScanReferences` if `no_reference_type` is set to `false`? 
* Sentinels are supposed to prevent the current bucket being drained and prevent the next bucket being opened. However, this work packet is in `FinalRefClosure` which happens after `SoftRefClosure`. So it means `SoftRefClosure` was already drained and the next bucket was already activated. What does it mean by adding another packet as the sentinel here? It sounds like this should just be a normal packet in `SoftRefClosure`.

Reply:
> Do we need this ReScanReferences if `no_reference_type` is set to false?

If `no_reference_type` is set to true, we can omit that because all references would be treated as strong references.  If false, we will need that.  I'll make the packet added conditionally according to the value of `no_reference_type`.

> ... However, this work packet is in FinalRefClosure which happens after SoftRefClosure. ...

That's a typo.  I meant to add the work packet to `FinalRefClosure`.  Soft and weak references should be re-scanned after all objects reachable from resurrected objects are traced, because some `SoftReference` and `WeakReference` objects will be resurrected, too.